packages/hierarchy-builder/src/hierarchy-builder/internal/operators/grouping/BaseClassGrouping.ts

```diff
@@ -0,0 +1,116 @@
+    }
+  }
+  const baseECClassesSorted = await sortByBaseClass(baseEntityAndRelationshipECClassesArray);
+  return baseECClassesSorted;
+}
+
+export async function createBaseClassGroupsForSingleBaseClass(
+  metadata: IMetadataProvider,
+  nodes: HierarchyNode[],
+  baseECClass: ECClass,
+): Promise<GroupingHandlerReturn> {
+  const finalAllNodeHierarchy = new Array<HierarchyNode>();
+  const finalGroupedNodeHierarchy = new Array<HierarchyNode>();
+  const baseClassGroupingNode: HierarchyNode = {
+    label: baseECClass.fullName,
+    key: {
+      type: "base-class-grouping",
+      class: { name: baseECClass.fullName, label: baseECClass.label ?? baseECClass.name },
+    },
+    children: [],
+  };
+  finalAllNodeHierarchy.push(baseClassGroupingNode);
+  finalGroupedNodeHierarchy.push(baseClassGroupingNode);
+  for (const node of nodes) {
+    if (HierarchyNode.isInstancesNode(node) && node.params?.grouping?.byBaseClasses) {
+      let classNameIsInBaseClassInfo = false;
+      // check if the node should be grouped by this baseClass
+      for (const classInfo of node.params.grouping.byBaseClasses.baseClassInfo) {
+        const specificClassName = `${classInfo.schemaName}.${classInfo.className}`;
+        if (specificClassName === baseECClass.fullName) {
+          classNameIsInBaseClassInfo = true;
+          break;
+        }
+      }
+      if (classNameIsInBaseClassInfo) {
+        const fullCurrentNodeClassName = node.key.instanceKeys[0].className;
+        const currentNodeECClass = await getClass(metadata, fullCurrentNodeClassName);
+        if (await currentNodeECClass.is(baseECClass)) {
+          if (finalAllNodeHierarchy.length > 0 && Array.isArray(baseClassGroupingNode.children)) {
+            baseClassGroupingNode.children.push(node);
+            continue;
+          }
+        }
+      }
+    }
+    finalAllNodeHierarchy.push(node);
+  }
+
+  // remove grouping node if it did not have any children
+  if (Array.isArray(baseClassGroupingNode.children) && baseClassGroupingNode.children.length === 0) {
+    finalAllNodeHierarchy.splice(0, 1);
+    finalGroupedNodeHierarchy.splice(0, 1);
+  }
+  return { allNodes: finalAllNodeHierarchy, groupedNodes: finalGroupedNodeHierarchy };
+}
+
+export function getAllBaseClasses(nodes: HierarchyNode[]): Set<string> {
+  const baseClasses = new Set<string>();
+  for (const node of nodes) {
+    if (HierarchyNode.isInstancesNode(node) && node.params?.grouping?.byBaseClasses) {
+      for (const classInfo of node.params.grouping.byBaseClasses.baseClassInfo) {
+        const specificClassName = `${classInfo.schemaName}.${classInfo.className}`;
+        baseClasses.add(specificClassName);
+      }
+    }
+  }
+  return baseClasses;
+}
+
+export async function sortByBaseClass(classes: ECClass[]) {
+  const output: ECClass[] = [];
+  const originalAmountOfClasses = classes.length;
+  while (output.length < originalAmountOfClasses) {
```

Comment:
looks like this could be optimized to at least O(n^2)

Reply:
Changed